pkg/adaptor/hypervisor/vsphere/vsphere.go

```diff
@@ -0,0 +1,275 @@
+		if err != nil {
+			return nil, err
+		}
+		datastoreref := types.NewReference(datastore.Reference())
+		relocateSpec.Datastore = datastoreref
+	}
+
+	cloneSpec := &types.VirtualMachineCloneSpec{
+		Location: relocateSpec,
+		PowerOn:  true,
+		Template: false,
+		Config:   &configSpec,
+	}
+
+	task, err := vm.Clone(ctx, vmfolder, vmname, *cloneSpec)
+	if err != nil {
+		logger.Printf("Can't clone to vm %s: %s", vmname, err)
+		return nil, err
+	}
+
+	info, err := task.WaitForResult(ctx, nil) // TODO Fix to have a timeout
+	if err != nil {
+		logger.Printf("wait for clone task failed")
+		return nil, err
+	}
+
+	clone := object.NewVirtualMachine(vim25Client, info.Result.(types.ManagedObjectReference))
+	name, err := clone.ObjectName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Printf("clone %s, Clone UUID %s created", name, clone.UUID(ctx))
+
+	//uuid := strings.Replace(mvm.Config.Uuid, "-", "", -1)
+
+	podNodeIPs, err := getIPs(clone) // TODO Fix to get all ips
+	if err != nil {
+		logger.Printf("failed to get IPs for the instance : %v ", err)
+		return nil, err
+	}
+
+	return &createInstanceOutput{
+		uuid: clone.UUID(ctx),
+		ips:  podNodeIPs,
+	}, nil
+}
+
+func getIPs(vm *object.VirtualMachine) ([]net.IP, error) { // TODO Fix to get all ips
+	var podNodeIPs []net.IP
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(300*time.Second))
+	defer cancel()
+
+	logger.Printf("start waiting for cloned vm ip")
+	ip, err := vm.WaitForIP(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Printf("VM IP=%s", ip)
+	ip_item := net.ParseIP(ip)
+	if ip_item == nil {
+		return nil, fmt.Errorf("failed to parse pod node IP %q", ip)
+	}
+	podNodeIPs = append(podNodeIPs, ip_item)
+
+	return podNodeIPs, nil
+}
+
+func DeleteInstance(ctx context.Context, vim25Client *vim25.Client, vmcfg *Config, vmname string) (err error) {
+
+	var (
+		task  *object.Task
+		state types.VirtualMachinePowerState
+	)
+
+	finder := find.NewFinder(vim25Client)
+
+	// If vmcfg.Datacenter is null DatacenterOrDefault will return the default datacenter
+	dc, err := finder.DatacenterOrDefault(ctx, vmcfg.Datacenter)
+
+	if err != nil {
+		logger.Printf("can't get vcenter datacenter %s", vmcfg.Datacenter)
+		return err
+	}
+
+	finder.SetDatacenter(dc)
+
+	vm_path := path.Join(dc.InventoryPath, "vm", vmcfg.Deployfolder, vmname)
+
+	vm, err := finder.VirtualMachine(ctx, vm_path) //TODO may need find by UUID
+	if err != nil {
+		logger.Printf("can't find VM %s to delete it", vm_path)
+		return err
+	}
+
+	state, err = vm.PowerState(ctx)
+	if err != nil {
+		return err
+	}
+
+	if state == types.VirtualMachinePowerStatePoweredOn {
+		task, err = vm.PowerOff(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Ignore error since the VM may already been in powered off state.
+		// vm.Destroy will fail if the VM is still powered on.
+		_ = task.Wait(ctx)
+	}
+
+	task, err = vm.Destroy(ctx)
+	if err != nil {
+		return err
+	}
+
+	task.Wait(ctx)
```

Comment:
@caavery is this needed ?

Reply:
In all the vsphere go code I've looked at it is common practice to wait for the destroy to complete

Reply:
So would this fix it? 
_ = task.Wait(ctx)

Reply:
Yeah you can use the above